README.md
# Data-Lifecycle-Manager component

This component is part of the 5G-MEDIA MAPE service. Take a look in the [MAPE](https://github.com/5g-media/mape) repository.

## Introduction

This component:
- consumes messages from the `ns` topic (intra-OSM kafka bus, OSM release>3)
- upon NS `terminate` event, remove the existing VDU entries from the Redis
- upon NS `terminated` event, remove the existing series from the InfluxDB database considering the tag `ns_uuid` (disabled by default)

The OSM release 4/5 publishes messages in the topic `ns` using a set of keys such as:
- [x] instantiate
- [x] instantiated
- [x] terminate
- [x] terminated

## Requirements
- Python 3.5+ 
   + a set of python packages are used (see requirements.txt).
- The Apache Kafka broker must be accessible from the component.
- The InfluxDB must be accessible from the component.
- The Redis must be accessible from the component.


## Installation/Deployment

To build the docker image, copy the bash script included in the `bash_scripts/`folder in the parent folder of the project and then, run:
```bash
   chmod +x build_docker_image.sh
   ./build_docker_image.sh
```

Considering the availability of the docker image, you can deploy the component as follows:
```bash
$ sudo docker run -p 8882:3333 --name data_lcm --restart always \
    -e PUBLIC_IP="kafka_ivp4" \
    -e OSM_IP="192.168.1.196" \
    -e OSM_KAFKA_IP="192.168.1.175" \
    -e OSM_KAFKA_PORT="9094" \
    -e OSM_USER="admin" \
    -e OSM_PWD="password" \
    -e REDIS_PORT="6379" \
    -e REDIS_EXPIRATION_SEC="14400" \
    -e KAFKA_PORT="9092" \
    -e INFLUXDB_DB_NAME="monitoring" \
    -e INFLUXDB_USER="root" \
    -e INFLUXDB_PWD="password" \
    -e INFLUXDB_PORT="8086" \
    -dit ns_manager
```

## Usage

Access the docker container:
```bash
$ sudo docker exec -it data_lcm bash
```

Start the NS manager service through the supervisor:
```bash
$ service supervisor start && supervisorctl start data-lcm
```

Stop the service through the supervisor:
```bash
$ supervisorctl stop data-lcm
```

Stop the supervisor service:
```bash
$ service supervisor stop 
```

You are able to see to status of service from browser using the supervisor UI;
type the URL http://{ipv4}:8882.

## Tests
Access the Kafka cli in the running kafka container:
```bash
sudo docker exec -it kafka-docker_kafka_1 bash
```

Run the bash script to produce a message using key:
```bash
$ bin/kafka-console-producer.sh --broker-list <intra-OSM-kafka-ipv4>:9094 --topic ns --property "parse.key=true" --property "key.separator=:"
```

The message should include a key and a value in the format: `key:value`. Check the messages under the folder `samples`.
Consider that these messages are in `JSON` (dict) while the actual one must be `YAML`.

To inspect the keys in the redis, follow the steps assuming that no password is needed:
```bash
$ sudo docker exec -it redis-cache bash
...
# redis-cli -n 0
...
> keys "ocata-devstack:*" // list keys having a pattern
1) "ocata-devstack:9c1bb4ad-1481-4d0f-be7f-a1928a014dfa"
2) "ocata-devstack:6e761377-fad3-46ec-a057-6e82e804d4d5"
3) "ocata-devstack:aa21a181-c3b1-4318-b0f3-df6f2c09b24b"
4) "ocata-devstack:69f9d7d7-c3f0-4cee-b0bd-4113d8c2ac43"
5) "ocata-devstack:47e43530-8a9e-4539-a54c-41dace08193b"
6) "ocata-devstack:65b297f2-de6c-45da-9a1b-8058efcba169"

...
> get "<key>" // get value by given key
"{\"vim\": {\"url\": \"http://192.168.1.147/identity/v3\", \"type\": \"openstack\", \"name\": \"ocata-devstack\", \"uuid\": \"41dab0c0-35f4-4c40-b1cd-13e4a79dab48\"}, \"vnf\": {\"vnfd_id\": \"16c40d2e-7a1b-4f22-9e50-3f7ede3e9fc4\", \"id\": \"a1e94c24-1155-400d-9d28-57ed00da7ea2\"}, \"ns\": {\"nsd_name\": \"cirros_2vnf_ns\", \"name\": \"takis-0007\", \"id\": \"47546676-fbb2-4990-bb3e-ca0b188f0072\", \"nsd_id\": \"d5c99561-ec46-4480-8377-b5b218b8b1e5\"}, \"vdu\": {\"mgmt-interface\": null, \"status\": \"ACTIVE\",  \"name\": \"takis-0007-1-cirros_vnfd-VM-1\", \"id\": \"65b297f2-de6c-45da-9a1b-8058efcba169\", \"ip_address\": \"192.168.157.1\"}}"
...

> mget "<key1>" "<key2>" // get values by given multiple keys
1) "{\"vim\": {\"url\": \"http://192.168.1.147/identity/v3\", \"type\": \"openstack\", \"name\": \"ocata-devstack\", \"uuid\": \"41dab0c0-35f4-4c40-b1cd-13e4a79dab48\"}, \"vnf\": {\"vnfd_id\": \"16c40d2e-7a1b-4f22-9e50-3f7ede3e9fc4\", \"id\": \"a1e94c24-1155-400d-9d28-57ed00da7ea2\"}, \"ns\": {\"nsd_name\": \"cirros_2vnf_ns\", \"name\": \"takis-0007\", \"id\": \"47546676-fbb2-4990-bb3e-ca0b188f0072\", \"nsd_id\": \"d5c99561-ec46-4480-8377-b5b218b8b1e5\"}, \"vdu\": {\"mgmt-interface\": null, \"status\": \"ACTIVE\", \"name\": \"takis-0007-1-cirros_vnfd-VM-1\", \"id\": \"65b297f2-de6c-45da-9a1b-8058efcba169\", \"ip_address\": \"192.168.157.1\"}}"
2) "{\"vnf\": {\"id\": \"5ad8628f-7493-4e92-87e2-5e8646d37a44\", \"vnfd_id\": \"16c40d2e-7a1b-4f22-9e50-3f7ede3e9fc4\"}, \"ns\": {\"nsd_id\": \"d5c99561-ec46-4480-8377-b5b218b8b1e5\", \"id\": \"c19a1f29-4445-45e4-9700-496501876237\", \"nsd_name\": \"cirros_2vnf_ns\", \"name\": \"takis-0006\"}, \"vdu\": {\"status\": \"ACTIVE\", \"ip_address\": \"192.168.49.9\", \"id\": \"47e43530-8a9e-4539-a54c-41dace08193b\", \"mgmt-interface\": null, \"name\": \"takis-0006-2-cirros_vnfd-VM-1\"}, \"vim\": {\"url\": \"http://192.168.1.147/identity/v3\", \"type\": \"openstack\", \"name\": \"ocata-devstack\"}}"
...

> del "<key>" // delete entry by given key
(integer) 1
```
More info about the redis-cli is available [here](https://redis.io/topics/rediscli).

## Authors
- Singular Logic <pathanasoulis@ep.singularlogic.eu>

## Contributors
 - Contact with Authors
 
## Acknowledgements
This project has received funding from the European Union’s Horizon 2020 research and innovation programme under grant agreement No 761699. The dissemination of results herein reflects only the author’s view and the European Commission is not responsible for any use that may be made of the information it contains.

## License
[Apache 2.0](LICENSE.md)




CHANGELOG.md
# CHANGELOG
---

## 2018-11-28
- Use redis for caching. This service listens the NS events from the intra-OSM-r4 kafka bus (topic `ns`)
and based on the type of the event adds an entry in redis or deletes an existing entry from the redis.

## 2018-07-25
- Delete the series related to a NS UUID from the InfluxDB (on demand) 



utils.py
from exceptions import NsValueIsNotDict, NsUuidDoesNotExist
from nbiapi.identity import bearer_token
from nbiapi.ns import Ns
from nbiapi.vim import Vim
from nbiapi.vnf import Vnf
from settings import OSM_ADMIN_CREDENTIALS


def get_ns_uuid_from_message(message):
    """Extract the UUID of the NS to be instantiated/terminated from the message

    In the OSM  r4, the message looks like as follows:
    --
    {
        "_admin": {
            "created": "1527159237.137106",
            "modified": "1527159237.137106"
        },
        "_id": "f850ebf3-a487-4fe6-bafa-6245c8b537e1",
        "id": "f850ebf3-a487-4fe6-bafa-6245c8b537e1",
        "isAutomaticInvocation": "false",
        "isCancelPending": "false",
        "lcmOperationType": "terminate",
        "links": {
            "nsInstance": "/osm/nslcm/v1/ns_instances/af6de0c1-7279-427e-9b68-1fa0e493d31d",
            "self": "/osm/nslcm/v1/ns_lcm_op_occs/f850ebf3-a487-4fe6-bafa-6245c8b537e1"
        },
        "nsInstanceId": "af6de0c1-7279-427e-9b68-1fa0e493d31d",
        "operationParams": {
            "_id": "f850ebf3-a487-4fe6-bafa-6245c8b537e1",
            "autoremove": "true",
            "nsInstanceId": "af6de0c1-7279-427e-9b68-1fa0e493d31d"
        },
        "operationState": "PROCESSING",
        "startTime": "1527159237.1370444",
        "statusEnteredTime": "1527159237.1370444"
    }
    --

    Args:
        message (dict): The value of the message as it was published in teh Kafka broker

    Returns:
        str: The UUID of the NS to be instantiated/terminated
    """
    if not isinstance(message, dict):
        raise NsValueIsNotDict('The value of the message in the ns topic is not dict'.format(message))
    if 'nsInstanceId' not in message:
        raise NsUuidDoesNotExist(
            'The nsInstanceId key is not included in the value of the message {} in the topic ns'.format(message))
    return str(message['nsInstanceId'])


def convert_byte_to_str(term):
    """Convert a term from the bytes to str

    Args:
        term (bytes): The term in bytes

    Returns:
        str: The term in str
    """
    return term.decode("utf-8")


def compose_redis_key(vim_name, vdu_uuid):
    """Compose the key for redis given vim name and vdu uuid

    Args:
        vim_name (str): The VIM name
        vdu_uuid (str): The VDU uuid (NFVI based)

    Returns:
        str: the key for redis
    """
    return "{}:{}".format(vim_name.lower(), vdu_uuid)


def get_vim_info(vim_uuid=None):
    """Get the VIM name, type, url by given VIM uuid

    Args:
        vim_uuid (str): The VIM uuid

    Returns:
        dict: the VIM uuid, name, type and url
    """
    if vim_uuid is None:
        return {"uuid": vim_uuid, "name": None, "type": None, "url": None}

    token = bearer_token(OSM_ADMIN_CREDENTIALS.get('username'), OSM_ADMIN_CREDENTIALS.get('username'))
    vim = Vim(token)
    response = vim.get(vim_uuid=vim_uuid)
    data = response.json()
    vim_info = {
        "uuid": vim_uuid,
        "name": data.get('name', None),
        "type": data.get('vim_type', None),
        "url": data.get('vim_url', None)
    }
    return vim_info


def get_vdus_info(ns_uuid=None):
    """Get information about NS, VNF(s) and VDU(s) by given NS uuid

    Args:
        ns_uuid (str): The NS uuid

    Returns:
        dict: ns, vnf and vdu info
    """
    vdus_info = []
    if ns_uuid is None:
        return vdus_info

    token = bearer_token(OSM_ADMIN_CREDENTIALS.get('username'), OSM_ADMIN_CREDENTIALS.get('username'))
    ns = Ns(token)
    ns_response = ns.get(ns_uuid=ns_uuid)
    nsr = ns_response.json()

    # Get Vim
    vim_uuid = nsr.get('datacenter', None)
    vim_info = get_vim_info(vim_uuid=vim_uuid)

    # Get the Vnf UUIDs, members of the NS
    vnf_uuids = nsr.get('constituent-vnfr-ref', [])

    vnfr = Vnf(token)

    for vnf_uuid in vnf_uuids:
        vnf_response = vnfr.get(vnf_uuid=vnf_uuid)
        vnf_record = vnf_response.json()

        # VDUs info
        vdu_records = vnf_record.get('vdur', [])

        for vdu_record in vdu_records:
            mano = {
                "vim": vim_info,
                "ns": {
                    "id": ns_uuid,
                    "name": nsr.get('name-ref', None),
                    "nsd_id": nsr.get('nsdId', None),
                    "nsd_name": nsr.get('nsd-name-ref', None)
                },
                "vnf": {
                    "id": vnf_record.get("id", None),
                    "name": None, # not provided in osm r4
                    "short_name": None, # not provided in osm r4
                    "vnfd_id": vnf_record.get("vnfd-id", None),
                    "vnfd_name": None # not provided in osm r4
                },
                "vdu": {
                    "id": vdu_record.get("vim-id", None),  # NFVI-based uuid
                    "image_id": None,
                    "flavor": {},
                    "status": vdu_record.get("status", None),
                    "ip_address": vdu_record.get("ip-address", None),
                    "mgmt-interface": None  # future usage
                }
            }
            vdus_info.append(mano)
    return vdus_info


daemon.py
import redis
import json
import yaml
import logging.config
from kafka import KafkaConsumer
from influxdb import InfluxDBClient
from utils import convert_byte_to_str, get_vdus_info, compose_redis_key
from exceptions import NsValueIsNotDict, NsUuidDoesNotExist
from settings import KAFKA_SERVER, KAFKA_CLIENT_ID, KAFKA_API_VERSION, KAFKA_NS_MANAGER_TOPIC, INFLUX_DATABASES, \
    REDIS_HOST, REDIS_PORT, REDIS_NFVI_DB, REDIS_EXPIRATION_SECONDS, LOGGING

logging.config.dictConfig(LOGGING)
logger = logging.getLogger(__name__)


def main():
    # See more: https://kafka-python.readthedocs.io/en/master/apidoc/KafkaConsumer.html
    consumer = KafkaConsumer(bootstrap_servers=KAFKA_SERVER, client_id=KAFKA_CLIENT_ID, enable_auto_commit=True,
                             value_deserializer=lambda v: yaml.safe_load(v.decode('utf-8', 'ignore')),
                             api_version=KAFKA_API_VERSION, )
    consumer.subscribe(KAFKA_NS_MANAGER_TOPIC)

    # See more: http://influxdb-python.readthedocs.io/en/latest/examples.html
    influx_client = InfluxDBClient(host=INFLUX_DATABASES['default']['HOST'], port=INFLUX_DATABASES['default']['PORT'],
                                   username=INFLUX_DATABASES['default']['USERNAME'],
                                   password=INFLUX_DATABASES['default']['PASSWORD'],
                                   database=INFLUX_DATABASES['default']['NAME'])

    # See more: https://redis-py.readthedocs.io/en/latest/
    redis_conn = redis.Redis(host=REDIS_HOST, port=REDIS_PORT, db=REDIS_NFVI_DB)

    # Run each message in "ns" topic
    for message in consumer:
        logger.info(message)
        message_key = convert_byte_to_str(message.key)
        message_value = message.value

        if message_key == "instantiate":
            """
            An indicative sample of message is available in the file `samples/instantiate.json`.
            Since the NS instantiation is still in progress, we skip this info.
            """
            pass

        elif message_key == "instantiated":
            """
            When a new Network Service has successfully instantiated through OSM, a new entry is added in the Redis.
            The Redis is used as a cache to avoid continuous requests in OSM-r4 NBI API in the translation process.
            
            The translator could check if there is info in the redis related to the metric that is under process.
            If there is (`hit` case), then no call in OSM r4 NBI API is needed. In `miss` case, the OSM will be used. 
            
            The key is a composition of <vim_name>:<vdu_uuid>. Use lower function in `vim_name`.
            An indicative sample of message is available in the file `samples/instantiated.json`.
            """
            pass

            # # Get the operational state of the process
            # ns_operation_state = message_value['operationState']
            # if ns_operation_state != 'COMPLETED':
            #     continue
            # ns_uuid = message_value.get('nsr_id', None)
            #
            # # Find the vdus for the new Network Service
            # vdu_records = get_vdus_info(ns_uuid=ns_uuid)
            #
            # # Update the entries in the redis, if needed
            # for vdu_record in vdu_records:
            #     vim_name = vdu_record.get("vim", {}).get('name', None)
            #     if vim_name is None:
            #         continue
            #
            #     vdu_uuid = vdu_record.get("vdu", {}).get('id', None)
            #     if vdu_uuid is None:
            #         continue
            #
            #     # Compose the redis key
            #     redis_key = compose_redis_key(vim_name, vdu_uuid)
            #
            #     # Check if there is value for the given key. If no, insert a value. Otherwise, do nothing.
            #     existing_value = redis_conn.get(name=redis_key)
            #     if existing_value is None:
            #         redis_conn.set(name=redis_key, value=json.dumps(vdu_record), ex=REDIS_EXPIRATION_SECONDS)
            #         logger.info("[Redis] Add key: `{}` with value: `{}`".format(redis_key, vdu_record))

        elif message_key == "terminate":
            """
            In this step, the NS termination is still in progress. However, this the proper time to remove the entries
            from redis since we can retrieve the vim information, since we can invoke the ns-instance web service. 
            
            An indicative sample of message is available in the file `samples/terminate.json`
            """
            ns_uuid = message_value.get('nsInstanceId', None)
            if ns_uuid is None:
                continue

            # Remove the entries related to the `ns_uuid` in Redis.
            vdu_records = get_vdus_info(ns_uuid=ns_uuid)
            for vdu_record in vdu_records:
                vim_name = vdu_record.get("vim", {}).get('name', None)
                if vim_name is None:
                    continue
                vdu_uuid = vdu_record.get("vdu", {}).get('id', None)
                if vdu_uuid is None:
                    continue

                # Check if there are keys with the pattern: *:{vdu_uuid}.
                # If yes, delete the relevant entry. Otherwise, do nothing.
                vdu_keys = redis_conn.keys(pattern="*:{}".format(vdu_uuid))
                vtranscoder_keys = redis_conn.keys(pattern="*_metrics:*")
                spectator_keys = redis_conn.keys(pattern="spectators.vtranscoder3d.metrics:transcoder*")
                matched_keys = vdu_keys + vtranscoder_keys + spectator_keys
                for k in matched_keys:
                    existing_value = redis_conn.get(name=k)
                    if existing_value is not None:
                        redis_conn.delete(k)
                        logger.info("[Redis] Delete key: `{}`".format(k))

        elif message_key == "terminated":
            """           
            When a new Network Service is terminated through OSM, delete the data from `InfluxDB` and the Redis. 
            The key is a composition of <vim_name>:<vdu_uuid>.
            
            An indicative sample of message is available in the file `samples/terminated.json`
            """
            # Get the operational state of the process
            ns_operation_state = message_value['operationState']
            if ns_operation_state != 'COMPLETED':
                continue

            # Delete the relevant redis entries
            spectator_keys = redis_conn.keys(pattern="spectators.vtranscoder3d.metrics:transcoder*")
            spectator_vnf_keys = redis_conn.keys(pattern="spectators.vtranscoder3d.metrics:vnf*")
            matched_keys = spectator_keys + spectator_vnf_keys
            for k in matched_keys:
                existing_value = redis_conn.get(name=k)
                if existing_value is not None:
                    redis_conn.delete(k)
                    logger.info("[Redis] Delete key: `{}`".format(k))

            ns_uuid = message_value.get('nsr_id', None)
            if ns_uuid is None:
                continue

            # # Remove the series with the "ns_uuid" tag from the Influxdb
            # try:
            #     logger.info("[InfluxDB] Delete series with tag `ns_uuid`={}".format(ns_uuid))
            #     influx_client.delete_series(database=INFLUX_DATABASES['default']['NAME'], tags={'ns_uuid': ns_uuid})
            # except (NsValueIsNotDict, NsUuidDoesNotExist, Exception) as exc:
            #     logger.exception(exc)

        elif message_key == "action":
            "Future usage"
            pass
        elif message_key == "show":
            "Future usage"
            pass
        elif message_key == "deleted":
            "Future usage"
            pass


if __name__ == '__main__':
    main()


nbiapi/vim.py
from settings import OSM_COMPONENTS, LOGGING
from httpclient.client import Client
import logging.config
import urllib3

urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
logging.config.dictConfig(LOGGING)
logger = logging.getLogger(__name__)


class Vim(object):
    """Description of Project class"""

    def __init__(self, token):
        """Constructor of Project class"""
        self.__client = Client(verify_ssl_cert=False)
        self.bearer_token = token

    def get_list(self):
        """Get the list of the registered VIMs in OSM r4

        Returns:
            obj: a requests object

        Examples:
            >>> from nbiapi.identity import bearer_token
            >>> from nbiapi.vim import Vim
            >>> from settings import OSM_ADMIN_CREDENTIALS
            >>> token = bearer_token(OSM_ADMIN_CREDENTIALS.get('username'), OSM_ADMIN_CREDENTIALS.get('username'))
            >>> vim = Vim(token)
            >>> entries = vim.get_list()
            >>> print(entries.json())
        """
        endpoint = '{}/osm/admin/v1/vim_accounts'.format(OSM_COMPONENTS.get('NBI-API'))
        headers = {"Authorization": "Bearer {}".format(self.bearer_token), "Accept": "application/json"}
        response = self.__client.get(endpoint, headers)
        return response

    def get(self, vim_uuid=None):
        """Get details for a VIM in OSM r4 by given VIM UUID

        Returns:
            obj: a requests object

        Examples:
            >>> from nbiapi.identity import bearer_token
            >>> from nbiapi.vim import Vim
            >>> from settings import OSM_ADMIN_CREDENTIALS
            >>> token = bearer_token(OSM_ADMIN_CREDENTIALS.get('username'), OSM_ADMIN_CREDENTIALS.get('username'))
            >>> vim = Vim(token)
            >>> vim_account = vim.get("41dab0c0-35f4-4c40-b1cd-13e4a79dab48")
            >>> print(vim_account.json())
        """
        endpoint = '{}/osm/admin/v1/vim_accounts/{}'.format(OSM_COMPONENTS.get('NBI-API'), vim_uuid)
        headers = {"Authorization": "Bearer {}".format(self.bearer_token), "Accept": "application/json"}
        response = self.__client.get(endpoint, headers)
        return response


deployment/run.sh
#!/bin/bash

# set the variables in the supervisor environment
sed -i "s/ENV_PUBLIC_IP/$PUBLIC_IP/g" /etc/supervisor/supervisord.conf
sed -i "s/ENV_OSM_IP/$OSM_IP/g" /etc/supervisor/supervisord.conf
sed -i "s/ENV_OSM_USER/$OSM_USER/g" /etc/supervisor/supervisord.conf
sed -i "s/ENV_OSM_PWD/$OSM_PWD/g" /etc/supervisor/supervisord.conf
sed -i "s/ENV_OSM_KAFKA_IP/$OSM_KAFKA_IP/g" /etc/supervisor/supervisord.conf
sed -i "s/ENV_OSM_KAFKA_PORT/$OSM_KAFKA_PORT/g" /etc/supervisor/supervisord.conf
sed -i "s/ENV_REDIS_PORT/$REDIS_PORT/g" /etc/supervisor/supervisord.conf
sed -i "s/ENV_REDIS_EXPIRATION_SEC/$REDIS_EXPIRATION_SEC/g" /etc/supervisor/supervisord.conf
sed -i "s/ENV_INFLUXDB_DB_NAME/$INFLUXDB_DB_NAME/g" /etc/supervisor/supervisord.conf
sed -i "s/ENV_INFLUXDB_USER/$INFLUXDB_USER/g" /etc/supervisor/supervisord.conf
sed -i "s/ENV_INFLUXDB_PWD/$INFLUXDB_PWD/g" /etc/supervisor/supervisord.conf
sed -i "s/ENV_INFLUXDB_PORT/$INFLUXDB_PORT/g" /etc/supervisor/supervisord.conf

# Restart services
service supervisor start && service supervisor status

# Makes services start on system start
update-rc.d supervisor defaults

echo "Initialization completed."
tail -f /dev/null  # Necessary in order for the container to not stop


settings.py
import os

PROJECT_ROOT = os.path.dirname(os.path.realpath(__file__))

# =================================
# KAFKA SETTINGS
# =================================
KAFKA_SERVER = "{}:{}".format(os.environ.get("OSM_KAFKA_IP", "10.100.176.66"),
                              os.environ.get("OSM_KAFKA_PORT", "9094"))
KAFKA_CLIENT_ID = 'data-lcm'
KAFKA_API_VERSION = (1, 1, 0)
KAFKA_NS_MANAGER_TOPIC = 'ns'  # the selected name is used from the OSM-r5

# =================================
# OSM SETTINGS
# =================================
OSM_IP = os.environ.get("OSM_IP", "10.100.176.66")
OSM_ADMIN_CREDENTIALS = {"username": os.environ.get("OSM_USER", "admin"),
                         "password": os.environ.get("OSM_PWD", "password")}
OSM_COMPONENTS = {"UI": 'http://{}:80'.format(OSM_IP),
                  "NBI-API": 'https://{}:9999'.format(OSM_IP),
                  "RO-API": 'http://{}:9090'.format(OSM_IP)}

# =================================
# INFLUXDB SETTINGS
# =================================
# See InfluxDBClient class
INFLUX_DATABASES = {
    'default': {
        'ENGINE': 'influxdb',
        'NAME': os.environ.get("INFLUXDB_DB_NAME", 'monitoring'),
        'USERNAME': os.environ.get("INFLUXDB_USER", 'root'),
        'PASSWORD': os.environ.get("INFLUXDB_PWD", 'password'),
        'HOST': os.environ.get("PUBLIC_IP", "192.168.1.175"),
        'PORT': os.environ.get("INFLUXDB_PORT", 8086)
    }
}

# =================================
# REDIS SETTINGS
# =================================
REDIS_HOST = os.environ.get("PUBLIC_IP", "10.100.176.70")
REDIS_PORT = os.environ.get("REDIS_PORT", 6379)
REDIS_NFVI_DB = 0
REDIS_PASSWORD = None
REDIS_EXPIRATION_SECONDS = os.environ.get("REDIS_EXPIRATION_SEC", 86400)  # default: 24 hours
# REDIS_SSL = False # default
# REDIS_SSL_KEYFILE = None # default
# REDIS_CERTFILE = None # default
# REDIS_SSL_CERT_REQS = u'required' # default
# REDIS_CA_CERTS = None # default


# ==================================
# LOGGING SETTINGS
# ==================================
# See more: https://docs.python.org/3.5/library/logging.config.html
LOGGING = {
    'version': 1,
    'disable_existing_loggers': True,
    'formatters': {
        'detailed': {
            'class': 'logging.Formatter',
            'format': "[%(asctime)s] - [%(name)s:%(lineno)s] - [%(levelname)s] %(message)s",
        },
        'simple': {
            'class': 'logging.Formatter',
            'format': '%(name)-15s %(levelname)-8s %(processName)-10s %(message)s'
        }
    },
    'handlers': {
        'console': {
            'class': 'logging.StreamHandler',
            'level': 'INFO',
            'formatter': 'simple',
        },
        'daemon': {
            'class': 'logging.handlers.RotatingFileHandler',
            'filename': "{}/logs/access.log".format(PROJECT_ROOT),
            'mode': 'w',
            'formatter': 'detailed',
            'level': 'INFO',
            'maxBytes': 2024 * 2024,
            'backupCount': 5,
        },
        'errors': {
            'class': 'logging.handlers.RotatingFileHandler',
            'filename': "{}/logs/error.log".format(PROJECT_ROOT),
            'mode': 'w',
            'level': 'ERROR',
            'formatter': 'detailed',
            'maxBytes': 2024 * 2024,
            'backupCount': 5,
        },
    },
    'loggers': {
        'daemon': {
            'handlers': ['daemon']
        },
        'errors': {
            'handlers': ['errors']
        }
    },
    'root': {
        'level': 'DEBUG',
        'handlers': ['console']
    },
}
